File: app/src/main/java/com/supercilex/robotscouter/ui/scout/AppBarViewHolderBase.java
package com.supercilex.robotscouter.ui.scout;

import android.arch.lifecycle.LifecycleFragment;
import android.arch.lifecycle.LiveData;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.os.Bundle;
import android.support.annotation.CallSuper;
import android.support.annotation.ColorInt;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.CollapsingToolbarLayout;
import android.support.v4.app.ActivityCompat;
import android.support.v4.app.Fragment;
import android.support.v7.graphics.Palette;
import android.support.v7.widget.Toolbar;
import android.text.TextUtils;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.ImageView;
import android.widget.ProgressBar;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.DataSource;
import com.bumptech.glide.load.engine.GlideException;
import com.bumptech.glide.request.RequestListener;
import com.bumptech.glide.request.RequestOptions;
import com.bumptech.glide.request.target.Target;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.gms.tasks.Task;
import com.google.android.gms.tasks.TaskCompletionSource;
import com.google.android.gms.tasks.Tasks;
import com.supercilex.robotscouter.R;
import com.supercilex.robotscouter.data.util.TeamHelper;
import com.supercilex.robotscouter.ui.ShouldUploadMediaToTbaDialog;
import com.supercilex.robotscouter.ui.TeamMediaCreator;

public abstract class AppBarViewHolderBase
        implements OnSuccessListener<Void>, View.OnLongClickListener, TeamMediaCreator.StartCaptureListener, ActivityCompat.OnRequestPermissionsResultCallback {
    private final Fragment mFragment;
    protected final Toolbar mToolbar;
    protected final CollapsingToolbarLayout mHeader;
    private final ImageView mBackdrop;
    private final ProgressBar mMediaLoadProgress;

    private final TaskCompletionSource<Void> mOnMenuReadyTask = new TaskCompletionSource<>();
    private final Task mOnScoutingReadyTask;

    protected TeamHelper mTeamHelper;

    private TeamMediaCreator mMediaCapture;
    private final OnSuccessListener<TeamHelper> mMediaCaptureListener = teamHelper -> {
        mTeamHelper.copyMediaInfo(teamHelper);
        mTeamHelper.forceUpdateTeam();
    };

    private MenuItem mNewScoutItem;
    private MenuItem mAddMediaItem;
    private MenuItem mVisitTeamWebsiteItem;
    private MenuItem mDeleteScoutItem;
    private boolean mIsDeleteScoutItemVisible;

    protected AppBarViewHolderBase(LiveData<TeamHelper> listener,
                                   LifecycleFragment fragment,
                                   View rootView,
                                   Task onScoutingReadyTask) {
        mFragment = fragment;
        mToolbar = rootView.findViewById(R.id.toolbar);
        mHeader = rootView.findViewById(R.id.header);
        mBackdrop = rootView.findViewById(R.id.backdrop);
        mMediaLoadProgress = rootView.findViewById(R.id.progress);
        mOnScoutingReadyTask = onScoutingReadyTask;
        mMediaCapture = TeamMediaCreator.newInstance(mFragment, mTeamHelper, mMediaCaptureListener);

        mBackdrop.setOnLongClickListener(this);
        mTeamHelper = listener.getValue();
        listener.observe(fragment, helper -> {
            if (helper == null) return;
            mTeamHelper = helper;
            bind();
        });
    }

    @CallSuper
    protected void bind() {
        mToolbar.setTitle(mTeamHelper.toString());
        mMediaCapture.setTeamHelper(mTeamHelper);
        loadImages();
        bindMenu();
    }

    private void loadImages() {
        mMediaLoadProgress.setVisibility(View.VISIBLE);

        String media = mTeamHelper.getTeam().getMedia();
        Glide.with(mBackdrop)
                .asBitmap()
                .load(media)
                .apply(RequestOptions.centerCropTransform().error(R.drawable.ic_memory_grey_48dp))
                .listener(new RequestListener<Bitmap>() {
                    @Override
                    public boolean onResourceReady(Bitmap resource,
                                                   Object model,
                                                   Target<Bitmap> target,
                                                   DataSource dataSource,
                                                   boolean isFirstResource) {
                        mMediaLoadProgress.setVisibility(View.GONE);

                        if (resource != null && !resource.isRecycled()) {
                            Palette.from(resource).generate(palette -> {
                                Palette.Swatch vibrantSwatch = palette.getVibrantSwatch();
                                if (vibrantSwatch != null) {
                                    updateScrim(vibrantSwatch.getRgb(), resource);
                                    return;
                                }

                                Palette.Swatch dominantSwatch = palette.getDominantSwatch();
                                if (dominantSwatch != null) {
                                    updateScrim(dominantSwatch.getRgb(), resource);
                                }
                            });
                        }
                        return false;
                    }

                    @Override
                    public boolean onLoadFailed(@Nullable GlideException e,
                                                Object model,
                                                Target<Bitmap> target,
                                                boolean isFirstResource) {
                        mMediaLoadProgress.setVisibility(View.GONE);
                        return false;
                    }
                })
                .into(mBackdrop);
    }

    @CallSuper
    protected void updateScrim(@ColorInt int color, Bitmap bitmap) {
        mHeader.setContentScrimColor(getTransparentColor(color));
    }

    public final void initMenu(Menu menu) {
        mNewScoutItem = menu.findItem(R.id.action_new_scout);
        mAddMediaItem = menu.findItem(R.id.action_add_media);
        mVisitTeamWebsiteItem = menu.findItem(R.id.action_visit_team_website);
        mDeleteScoutItem = menu.findItem(R.id.action_delete);

        menu.findItem(R.id.action_visit_tba_website)
                .setTitle(mFragment.getString(R.string.visit_team_website_on_tba,
                                              mTeamHelper.getTeam().getNumber()));
        mVisitTeamWebsiteItem.setTitle(mFragment.getString(R.string.visit_team_website,
                                                           mTeamHelper.getTeam().getNumber()));
        if (!mOnScoutingReadyTask.isComplete()) mNewScoutItem.setVisible(false);

        mOnMenuReadyTask.trySetResult(null);
        bindMenu();
    }

    public void setDeleteScoutMenuItemVisible(boolean visible) {
        mIsDeleteScoutItemVisible = visible;
        bindMenu();
    }

    private void bindMenu() {
        Tasks.whenAll(mOnMenuReadyTask.getTask(), mOnScoutingReadyTask).addOnSuccessListener(this);
    }

    @Override
    public final void onSuccess(Void aVoid) {
        mNewScoutItem.setVisible(true);
        mAddMediaItem.setVisible(mTeamHelper.isOutdatedMedia());
        mVisitTeamWebsiteItem.setVisible(!TextUtils.isEmpty(mTeamHelper.getTeam().getWebsite()));
        mDeleteScoutItem.setVisible(mIsDeleteScoutItemVisible);
    }

    private int getTransparentColor(@ColorInt int opaque) {
        return Color.argb(Math.round(Color.alpha(opaque) * 0.6f),
                          Color.red(opaque),
                          Color.green(opaque),
                          Color.blue(opaque));
    }

    @Override
    public boolean onLongClick(View v) {
        if (v.getId() == R.id.backdrop) {
            ShouldUploadMediaToTbaDialog.Companion.show(mFragment);
            return true;
        }
        return false;
    }

    @Override
    public void onStartCapture(boolean shouldUploadMediaToTba) {
        mMediaCapture.startCapture(shouldUploadMediaToTba);
    }

    @CallSuper
    public void onActivityResult(int requestCode, int resultCode) {
        mMediaCapture.onActivityResult(requestCode, resultCode);
    }

    @CallSuper
    @Override
    public void onRequestPermissionsResult(int requestCode,
                                           @NonNull String[] permissions,
                                           @NonNull int[] grantResults) {
        mMediaCapture.onRequestPermissionsResult(requestCode, permissions, grantResults);
    }

    @CallSuper
    public void onSaveInstanceState(Bundle outState) {
        outState.putAll(mMediaCapture.toBundle());
    }

    @CallSuper
    public void restoreState(Bundle savedInstanceState) {
        mMediaCapture = TeamMediaCreator.get(savedInstanceState, mFragment, mMediaCaptureListener);
    }
}

File: app/src/main/java/com/supercilex/robotscouter/ui/scouting/scout/AppBarViewHolderBase.kt
package com.supercilex.robotscouter.ui.scouting.scout

import android.arch.lifecycle.LifecycleFragment
import android.arch.lifecycle.LiveData
import android.arch.lifecycle.Observer
import android.graphics.Bitmap
import android.graphics.Color
import android.os.Bundle
import android.support.annotation.CallSuper
import android.support.annotation.ColorInt
import android.support.design.widget.CollapsingToolbarLayout
import android.support.v4.app.ActivityCompat
import android.support.v7.graphics.Palette
import android.support.v7.widget.Toolbar
import android.text.TextUtils
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.ImageView
import android.widget.ProgressBar
import com.bumptech.glide.Glide
import com.bumptech.glide.load.DataSource
import com.bumptech.glide.load.engine.GlideException
import com.bumptech.glide.request.RequestListener
import com.bumptech.glide.request.RequestOptions
import com.bumptech.glide.request.target.Target
import com.google.android.gms.tasks.OnSuccessListener
import com.google.android.gms.tasks.Task
import com.google.android.gms.tasks.TaskCompletionSource
import com.google.android.gms.tasks.Tasks
import com.supercilex.robotscouter.R
import com.supercilex.robotscouter.data.model.Team
import com.supercilex.robotscouter.ui.ShouldUploadMediaToTbaDialog
import com.supercilex.robotscouter.ui.TeamMediaCreator
import com.supercilex.robotscouter.util.data.model.copyMediaInfo
import com.supercilex.robotscouter.util.data.model.forceUpdateTeam
import com.supercilex.robotscouter.util.data.model.isOutdatedMedia

abstract class AppBarViewHolderBase protected constructor(private val fragment: LifecycleFragment,
                                                          rootView: View,
                                                          listener: LiveData<Team>,
                                                          private val onScoutingReadyTask: Task<*>) :
        OnSuccessListener<Void?>, View.OnLongClickListener,
        TeamMediaCreator.StartCaptureListener, ActivityCompat.OnRequestPermissionsResultCallback {
    protected var team: Team = listener.value!!

    protected val toolbar: Toolbar = rootView.findViewById(R.id.toolbar)
    protected val header: CollapsingToolbarLayout = rootView.findViewById(R.id.header)
    private val backdrop: ImageView = rootView.findViewById(R.id.backdrop)
    private val mediaLoadProgress: ProgressBar = rootView.findViewById(R.id.progress)

    private val mediaCaptureListener = OnSuccessListener<Team> {
        team.copyMediaInfo(it)
        team.forceUpdateTeam()
    }
    private var mediaCapture: TeamMediaCreator = TeamMediaCreator.newInstance(fragment, team, mediaCaptureListener)

    private val onMenuReadyTask = TaskCompletionSource<Nothing?>()
    private lateinit var newScoutItem: MenuItem
    private lateinit var addMediaItem: MenuItem
    private lateinit var visitTeamWebsiteItem: MenuItem
    private lateinit var deleteScoutItem: MenuItem
    private var isDeleteScoutItemVisible: Boolean = false

    init {
        backdrop.setOnLongClickListener(this)

        listener.observe(fragment, Observer {
            team = it ?: return@Observer
            bind()
        })
    }

    @CallSuper
    protected open fun bind() {
        toolbar.title = team.toString()
        mediaCapture.setTeam(team)
        loadImages()
        bindMenu()
    }

    private fun loadImages() {
        mediaLoadProgress.visibility = View.VISIBLE

        val media = team.media
        Glide.with(backdrop)
                .asBitmap()
                .load(media)
                .apply(RequestOptions.centerCropTransform().error(R.drawable.ic_memory_grey_48dp))
                .listener(object : RequestListener<Bitmap> {
                    override fun onResourceReady(resource: Bitmap?,
                                                 model: Any?,
                                                 target: Target<Bitmap>,
                                                 dataSource: DataSource,
                                                 isFirstResource: Boolean): Boolean {
                        mediaLoadProgress.visibility = View.GONE

                        if (resource != null && !resource.isRecycled) {
                            Palette.from(resource).generate { palette ->
                                palette.vibrantSwatch?.let {
                                    updateScrim(it.rgb, resource)
                                    return@generate
                                }

                                palette.dominantSwatch?.let { updateScrim(it.rgb, resource) }
                            }
                        }
                        return false
                    }

                    override fun onLoadFailed(e: GlideException?,
                                              model: Any?,
                                              target: Target<Bitmap>,
                                              isFirstResource: Boolean): Boolean {
                        mediaLoadProgress.visibility = View.GONE
                        return false
                    }
                })
                .into(backdrop)
    }

    @CallSuper
    protected open fun updateScrim(@ColorInt color: Int, bitmap: Bitmap?) =
            header.setContentScrimColor(getTransparentColor(color))

    fun initMenu(menu: Menu) {
        newScoutItem = menu.findItem(R.id.action_new_scout)
        addMediaItem = menu.findItem(R.id.action_add_media)
        visitTeamWebsiteItem = menu.findItem(R.id.action_visit_team_website)
        deleteScoutItem = menu.findItem(R.id.action_delete)

        menu.findItem(R.id.action_visit_tba_website).title =
                fragment.getString(R.string.visit_team_website_on_tba, team.number)
        visitTeamWebsiteItem.title = fragment.getString(R.string.visit_team_website, team.number)
        if (!onScoutingReadyTask.isComplete) newScoutItem.isVisible = false

        onMenuReadyTask.trySetResult(null)
        bindMenu()
    }

    fun setDeleteScoutMenuItemVisible(visible: Boolean) {
        isDeleteScoutItemVisible = visible
        bindMenu()
    }

    private fun bindMenu() =
            Tasks.whenAll(onMenuReadyTask.task, onScoutingReadyTask).addOnSuccessListener(this)

    override fun onSuccess(void: Void?) {
        newScoutItem.isVisible = true
        addMediaItem.isVisible = team.isOutdatedMedia
        visitTeamWebsiteItem.isVisible = !TextUtils.isEmpty(team.website)
        deleteScoutItem.isVisible = isDeleteScoutItemVisible
    }

    private fun getTransparentColor(@ColorInt opaque: Int): Int =
            Color.argb(Math.round(Color.alpha(opaque) * 0.6f),
                    Color.red(opaque),
                    Color.green(opaque),
                    Color.blue(opaque))

    override fun onLongClick(v: View): Boolean = if (v.id == R.id.backdrop) {
        ShouldUploadMediaToTbaDialog.show(fragment)
        true
    } else false

    override fun onStartCapture(shouldUploadMediaToTba: Boolean) =
            mediaCapture.startCapture(shouldUploadMediaToTba)

    @CallSuper
    fun onActivityResult(requestCode: Int, resultCode: Int) =
            mediaCapture.onActivityResult(requestCode, resultCode)

    @CallSuper
    override fun onRequestPermissionsResult(requestCode: Int,
                                            permissions: Array<String>,
                                            grantResults: IntArray) =
            mediaCapture.onRequestPermissionsResult(requestCode, permissions, grantResults)

    @CallSuper
    fun onSaveInstanceState(outState: Bundle) = outState.putAll(mediaCapture.toBundle())

    @CallSuper
    fun restoreState(savedInstanceState: Bundle) {
        mediaCapture = TeamMediaCreator.get(savedInstanceState, fragment, mediaCaptureListener)
    }
}
